### PR TITLE
[Ballista] Streaming style push-based shuffle and All-at-once stage scheduling in Ballista

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes [#1805.](https://github.com/apache/arrow-datafusion/issues/1805)

 # Rationale for this change

Add a new streaming style push based shuffle implementation.

# What changes are included in this PR?
1. new stream shuffle reader implementation
2. PushPartition gRpc call in Arrow-Flight
3. All-at-once stage scheduler

# Are there any user-facing changes?

No

### CODE REVIEW

File: ballista/rust/scheduler/src/planner.rs [hunk 1]
```diff
@@ -55,24 +56,41 @@ impl Default for DistributedPlanner {
...
-            job_id,
-            self.next_stage_id(),
-            new_plan,
-            None,
-        )?);
-        Ok(stages)
+        // re-plan the input execution plan and create All-at-once query stages.
+        // Now we just simply depends on the the stage count to decide whether to create All-at-once or normal stages.
+        // In future, we can have more sophisticated way to decide which way to go.
+        if stages.len() > 1 && stages.len() <= 4 {
```
Comment (thinkharderdev) [inline ballista/rust/scheduler/src/planner.rs:74]: If I understand the original design correctly, the "all-at-once" plan will only get scheduled when there are sufficient task slots available to run the entire plan. So should this be a function of the total number of partitions?

Comment (mingmwang) [reply ballista/rust/scheduler/src/planner.rs:74]: > If I understand the original design correctly, the "all-at-once" plan will only get scheduled when there are sufficient task slots available to run the entire plan. So should this be a function of the total number of partitions?

Yes, you are right. But currently the scheduler server doesn't have a clear view of how many task slots available. So here I just add simple check on the stage count. After @yahoNanJing refactor the scheduler state and keep more cpu/task info into the memory state, we can add more sophisticated check logic.

**Comment by alamb** [comment]
marking as draft (so it is easer to see what PRs are waiting for review)

**Comment by andygrove** [comment]
Closing this PR since it has not been updated in a long time. Feel free to re-open if this is still being worked on.

**Pull request closed without merging**